File: internal/sdk/connectors.go
package sdk

// Connectors resource (issue #54 CLI+SDK parity): the typed mirror of the
// connectors-framework HTTP surface in cmd/api/connectors.go.
//
// Endpoints (all under /v1/connectors, writes OWNER/ADMIN, reads MEMBER+):
//
//	POST   /connectors           -> CreateConnector
//	GET    /connectors           -> ListConnectors
//	GET    /connectors/{id}      -> GetConnector
//	DELETE /connectors/{id}      -> DeleteConnector
//	POST   /connectors/{id}/test -> TestConnector (live health check)
//
// SECURITY: secret VALUES never appear anywhere in this surface. SecretRef
// is a NAME reference into the secrets store and ConnectorConfig carries
// header TEMPLATES and auth-style parameters only — mirrored structurally so
// a value could not slip through the typed client.

import (
	"context"
	"time"
)

// ConnectorConfig is the connector's NON-SECRET request configuration.
type ConnectorConfig struct {
	// AuthStyle is one of none|bearer|basic|api_key_header ("" -> none).
	AuthStyle string `json:"auth_style,omitempty"`
	// Headers are static header templates merged into every built request.
	Headers map[string]string `json:"headers,omitempty"`
	// APIKeyHeader names the header for the api_key_header style.
	APIKeyHeader string `json:"api_key_header,omitempty"`
	// APIKeyPrefix is prepended to the resolved secret for api_key_header.
	APIKeyPrefix string `json:"api_key_prefix,omitempty"`
	// Username is the basic-auth username (the password comes from the
	// secret referenced by SecretRef).
	Username string `json:"username,omitempty"`
}

// Connector is the org-scoped registry row. LastCheckAt/LastCheckStatus
// carry the outcome of the most recent live health-check probe (nil/""
// while never checked — the handler emits last_check_at:null explicitly).
type Connector struct {
	ID              string          `json:"id"`
	Name            string          `json:"name"`
	Type            string          `json:"type"` // webhook | http
	BaseURL         string          `json:"base_url"`
	SecretRef       string          `json:"secret_ref"`
	Status          string          `json:"status"` // active | disabled
	Config          ConnectorConfig `json:"config"`
	CreatedBy       string          `json:"created_by"`
	CreatedAt       time.Time       `json:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at"`
	LastCheckAt     *time.Time      `json:"last_check_at"`
	LastCheckStatus string          `json:"last_check_status"` // ""|ok|error
}

// ConnectorList is the wrapped shape of GET /v1/connectors (name ASC).
type ConnectorList struct {
	Connectors []Connector `json:"connectors"`
}

// CreateConnectorRequest is the POST /v1/connectors body — exactly the
// handler's flattened payload (config fields inline, secret NAME only).
type CreateConnectorRequest struct {
	Name         string            `json:"name"`
	Type         string            `json:"type"`
	BaseURL      string            `json:"base_url"`
	AuthStyle    string            `json:"auth_style,omitempty"`
	Headers      map[string]string `json:"headers,omitempty"`
	APIKeyHeader string            `json:"api_key_header,omitempty"`
	APIKeyPrefix string            `json:"api_key_prefix,omitempty"`
	Username     string            `json:"username,omitempty"`
	SecretRef    string            `json:"secret_ref,omitempty"`
	Status       string            `json:"status,omitempty"`
}

// ConnectorTestResult is the outcome of one live health-check probe. Error
// messages never contain secret values (resolver failures surface the secret
// NAME and the underlying error kind only).
type ConnectorTestResult struct {
	ConnectorID string    `json:"connector_id"`
	Status      string    `json:"status"` // ok | error
	StatusCode  int       `json:"status_code"`
	LatencyMS   int64     `json:"latency_ms"`
	Error       string    `json:"error,omitempty"`
	CheckedAt   time.Time `json:"checked_at"`
}

// CreateConnector validates and registers a new org-scoped connector
// (POST /v1/connectors, 201). The request never carries secret VALUES.
func (c *Client) CreateConnector(ctx context.Context, req CreateConnectorRequest) (*Connector, error) {
	var out struct {
		Connector Connector `json:"connector"`
	}
	if err := c.do(ctx, httpMethodPost, "/connectors", nil, req, &out); err != nil {
		return nil, err
	}
	return &out.Connector, nil
}

// ListConnectors returns the caller's connectors, name ASC
// (GET /v1/connectors).
func (c *Client) ListConnectors(ctx context.Context) (*ConnectorList, error) {
	var out ConnectorList
	if err := c.do(ctx, httpMethodGet, "/connectors", nil, nil, &out); err != nil {
		return nil, err
	}
	if out.Connectors == nil {
		out.Connectors = []Connector{}
	}
	return &out, nil
}

// GetConnector returns one connector within the caller's organization
// (GET /v1/connectors/{id}).
func (c *Client) GetConnector(ctx context.Context, id string) (*Connector, error) {
	var out struct {
		Connector Connector `json:"connector"`
	}
	if err := c.do(ctx, httpMethodGet, "/connectors/"+urlPathEscape(id), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out.Connector, nil
}

// DeleteConnector hard-deletes one connector within the caller's
// organization (DELETE /v1/connectors/{id}; foreign/unknown ids surface as
// 404 without an existence leak).
func (c *Client) DeleteConnector(ctx context.Context, id string) error {
	var out struct {
		Deleted bool `json:"deleted"`
	}
	return c.do(ctx, httpMethodDelete, "/connectors/"+urlPathEscape(id), nil, nil, &out)
}

// TestConnector triggers the live health check (5s timeout; the outcome is
// recorded on the connector as last_check_at/last_check_status)
// (POST /v1/connectors/{id}/test).
func (c *Client) TestConnector(ctx context.Context, id string) (*ConnectorTestResult, error) {
	var out struct {
		Test ConnectorTestResult `json:"test"`
	}
	if err := c.do(ctx, httpMethodPost, "/connectors/"+urlPathEscape(id)+"/test", nil, nil, &out); err != nil {
		return nil, err
	}
	return &out.Test, nil
}

File: web/src/lib/api/connectors.ts
// Connectors resource (issue #30, cmd/api/connectors.go):
//
//   POST   /connectors            -> {"connector":{…}}   create (connectors.write — OWNER/ADMIN)
//   GET    /connectors            -> {"connectors":[…]}  list   (connectors.read — MEMBER+)
//   GET    /connectors/{id}       -> {"connector":{…}}   get    (connectors.read — MEMBER+)
//   DELETE /connectors/{id}       -> {"deleted":true}    delete (connectors.write — OWNER/ADMIN)
//   POST   /connectors/{id}/test  -> {"test":{…}}        live health check (connectors.write — OWNER/ADMIN)
//
// NOTE on "U" in CRUD: the connectors service has an Update method, but the
// HTTP surface exposes NO PUT/PATCH route (cmd/api/marketplace.go sibling
// audit, wave-5). Editing an existing connector is therefore not possible via
// the API and this client does not fake it — the create form doubles as a
// fresh-configuration entry point.
//
// SECURITY: secret VALUES never appear anywhere in this surface. secret_ref is
// a NAME reference into the secrets store; config carries header TEMPLATES and
// auth-style parameters only.

import { apiFetch } from './client'
import { asNumber, asString, pickField } from './types'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ConnectorConfig = {
  authStyle: string
  headers: Record<string, string>
  apiKeyHeader: string
  apiKeyPrefix: string
  username: string
}

export type Connector = {
  id: string
  name: string
  type: string
  baseUrl: string
  secretRef: string
  status: string
  config: ConnectorConfig
  createdBy: string
  lastCheckAt: string | null
  /** "" (never checked) | ok | error */
  lastCheckStatus: string
  createdAt?: string
  updatedAt?: string
}

export type ConnectorTestResult = {
  connectorId: string
  /** ok | error */
  status: string
  statusCode: number
  latencyMs: number
  error?: string
  checkedAt: string
}

export type CreateConnectorInput = {
  name: string
  /** webhook | http */
  type: string
  base_url: string
  /** none | bearer | basic | api_key_header */
  auth_style: string
  headers?: Record<string, string>
  api_key_header?: string
  api_key_prefix?: string
  username?: string
  secret_ref?: string
  /** active | disabled */
  status?: string
}

export const CONNECTOR_TYPES = ['http', 'webhook'] as const
export const CONNECTOR_AUTH_STYLES = ['none', 'bearer', 'basic', 'api_key_header'] as const
export const CONNECTOR_STATUSES = ['active', 'disabled'] as const

// ---------------------------------------------------------------------------
// Normalizers
// ---------------------------------------------------------------------------

function normalizeStringMap(value: unknown): Record<string, string> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return {}
  const out: Record<string, string> = {}
  for (const [key, raw] of Object.entries(value as Record<string, unknown>)) {
    const str = asString(raw)
    if (str !== undefined) out[key] = str
  }
  return out
}

function normalizeConnector(raw: unknown): Connector {
  const config = pickField(raw, 'config')
  const lastCheckAt = asString(pickField(raw, 'lastCheckAt', 'last_check_at'))
  return {
    id: asString(pickField(raw, 'id')) ?? '',
    name: asString(pickField(raw, 'name')) ?? 'Unnamed connector',
    type: (asString(pickField(raw, 'type')) ?? 'unknown').toLowerCase(),
    baseUrl: asString(pickField(raw, 'baseUrl', 'base_url')) ?? '',
    secretRef: asString(pickField(raw, 'secretRef', 'secret_ref')) ?? '',
    status: (asString(pickField(raw, 'status')) ?? 'unknown').toLowerCase(),
    config: {
      authStyle: (asString(pickField(config, 'authStyle', 'auth_style')) ?? 'none').toLowerCase(),
      headers: normalizeStringMap(pickField(config, 'headers')),
      apiKeyHeader: asString(pickField(config, 'apiKeyHeader', 'api_key_header')) ?? '',
      apiKeyPrefix: asString(pickField(config, 'apiKeyPrefix', 'api_key_prefix')) ?? '',
      username: asString(pickField(config, 'username')) ?? '',
    },
    createdBy: asString(pickField(raw, 'createdBy', 'created_by')) ?? '',
    lastCheckAt: lastCheckAt ?? null,
    lastCheckStatus: (asString(pickField(raw, 'lastCheckStatus', 'last_check_status')) ?? '').toLowerCase(),
    createdAt: asString(pickField(raw, 'createdAt', 'created_at')),
    updatedAt: asString(pickField(raw, 'updatedAt', 'updated_at')),
  }
}

function normalizeTestResult(raw: unknown): ConnectorTestResult {
  return {
    connectorId: asString(pickField(raw, 'connectorId', 'connector_id')) ?? '',
    status: (asString(pickField(raw, 'status')) ?? 'unknown').toLowerCase(),
    statusCode: asNumber(pickField(raw, 'statusCode', 'status_code')) ?? 0,
    latencyMs: asNumber(pickField(raw, 'latencyMs', 'latency_ms')) ?? 0,
    error: asString(pickField(raw, 'error')),
    checkedAt: asString(pickField(raw, 'checkedAt', 'checked_at')) ?? '',
  }
}

// ---------------------------------------------------------------------------
// Fetchers
// ---------------------------------------------------------------------------

function extractConnectorList(raw: unknown): Connector[] {
  const list = pickField(raw, 'connectors')
  return (Array.isArray(list) ? list : []).map(normalizeConnector)
}

export async function listConnectors(): Promise<Connector[]> {
  return extractConnectorList(await apiFetch<unknown>('/connectors'))
}

export async function getConnector(id: string): Promise<Connector> {
  const raw = await apiFetch<unknown>(`/connectors/${encodeURIComponent(id)}`)
  return normalizeConnector(pickField(raw, 'connector') ?? raw)
}

export async function createConnector(input: CreateConnectorInput): Promise<Connector> {
  const raw = await apiFetch<unknown>('/connectors', {
    method: 'POST',
    body: JSON.stringify(input),
  })
  return normalizeConnector(pickField(raw, 'connector') ?? raw)
}

export async function deleteConnector(id: string): Promise<void> {
  await apiFetch<unknown>(`/connectors/${encodeURIComponent(id)}`, { method: 'DELETE' })
}

/** Live health check; the outcome is also recorded on the connector row. */
export async function testConnector(id: string): Promise<ConnectorTestResult> {
  const raw = await apiFetch<unknown>(`/connectors/${encodeURIComponent(id)}/test`, { method: 'POST' })
  return normalizeTestResult(pickField(raw, 'test') ?? raw)
}
